internal/k3s/install.go

```diff
@@ -0,0 +1,221 @@
+		return err
+	}
+
+	fixed, err := dnsFixed()
+	if err != nil {
+		return err
+	}
+	c.dnsFixed = fixed
+
+	name, version, err := findBundle(c.BundlePath)
+	if err != nil {
+		return err
+	}
+
+	bundle := bundle.Tar(name)
+
+	fmt.Printf("Installing K3S %q\n", version)
+
+	err = errors.Join(
+		bundle.GetFiles(copyK3S, "k3s"),
+		bundle.GetFiles(copyAirgapImages, "k3s-airgap-*.tar"),
+		bundle.GetFiles(runInstallScript(ctx, c), "install.sh"),
+	)
+
+	if err != nil {
+		cleanup()
+		return err
+	}
+
+	if errors.Is(ctx.Err(), context.Canceled) {
+		cleanup()
+		return ctx.Err()
+	}
+
+	return nil
+}
+
+func cleanup() error {
+	//exec.Command("k3s-uninstall.sh").Run()
+	os.RemoveAll(k3sImagesPath)
+	os.Remove(k3sBinary)
+	os.Remove(k3sResolvConfPath)
+	return nil
+}
+
+func copyK3S(_ fs.FileInfo, r io.Reader) error {
+	f, err := os.OpenFile(k3sBinary, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, fs.FileMode(0755))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, r)
+	if err != nil && !errors.Is(err, io.EOF) {
+		err = errors.Join(err, f.Close())
+		err = errors.Join(err, os.Remove(k3sBinary))
+		return err
+	}
+	return nil
+}
+
+func copyAirgapImages(info fs.FileInfo, r io.Reader) error {
+	os.MkdirAll(k3sImagesPath, 0644)
+
+	f, err := os.OpenFile(path.Join(k3sImagesPath, info.Name()), os.O_CREATE|os.O_WRONLY, fs.FileMode(0644))
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(f, r)
+	if err != nil && !errors.Is(err, io.EOF) {
+		err = errors.Join(err, f.Close())
+		err = errors.Join(err, os.Remove(path.Join(k3sImagesPath, info.Name())))
+		return err
+	}
+	return nil
+}
+
+func runInstallScript(ctx context.Context, c InstallConfig) func(fs.FileInfo, io.Reader) error {
+	return func(fi fs.FileInfo, r io.Reader) error {
+		if c.Hostname != "" {
+			os.Setenv("K3S_HOSTNAME", c.Hostname)
+			os.Setenv("K3S_NODE_NAME", c.Hostname)
+		}
+		if c.dnsFixed {
+			os.Setenv("K3S_RESOLV_CONF", k3sResolvConfPath)
+		}
+		os.Setenv("INSTALL_K3S_BIN_DIR", k3SInstallDir)
+		os.Setenv("INSTALL_K3S_SKIP_DOWNLOAD", "true")
+		os.Setenv("INSTALL_K3S_SELINUX_WARN", "true")
+		os.Setenv("INSTALL_K3S_SKIP_SELINUX_RPM", "true")
+
+		k3sArgs := []string{
+			"--with-node-id",
+			fmt.Sprintf("--flannel-iface=%s", c.Iface),
+			fmt.Sprintf("--default-local-storage-path=%s", defaultLocalStoragePath),
+		}
+		if len(c.NodeIP) > 0 {
+			for _, ip := range c.NodeIP {
+				k3sArgs = append(k3sArgs, fmt.Sprintf("--node-ip=%s", ip))
+			}
+		}
+		if len(c.ExternalIP) > 0 {
+			for _, ip := range c.ExternalIP {
+				k3sArgs = append(k3sArgs, fmt.Sprintf("--node-external-ip=%s", ip))
+			}
+		}
+
+		os.Setenv("INSTALL_K3S_EXEC", strings.Join(k3sArgs, " "))
+
+		cmd := exec.CommandContext(ctx, "sh", "-")
+		cmd.Stdin = r
+
+		stdout, _ := cmd.StdoutPipe()
+		go io.Copy(os.Stdout, stdout)
+
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("install.sh: %w", err)
+		}
+
+		return nil
+	}
+}
+
+var resolvRegexp = regexp.MustCompile(`%s*nameserver.*`)
```

Comment:
minor: I'd better move dns manipulations into dns.go so that there would be less mixed up things